src/pdm/project/core.py

```diff
@@ -530,9 +530,12 @@ def make_self_candidate(self, editable: bool = True) -> Candidate:
         return make_candidate(req, name=self.meta.name, version=self.meta.version)
 
     def get_content_hash(self, algo: str = "md5") -> str:
-        # Only calculate sources and dependencies groups. Otherwise lock file is
+        # Generate a hash based on the external dependencies, as well
+        # as the identity of the project. Otherwise lock file is
         # considered as unchanged.
         dump_data = {
+            "version": self.meta.version,
```

Comment:
Do we have to count `version` in? That may change in EVERY commit if we are using dynamic versioning from SCM.

Reply:
Yes, I think so. Or we need to relax what version gets put into the lockfile for the local package.

I put up an isolated repro and observe the issue on 2.1.5, as well as on latest main before my fix. https://github.com/tgolsson/pdm-repros/tree/main/1470

Reply:
I don't think it is the correct approach to fixing the issue. The self package, ideally, shouldn't appear in the lockfile.

Reply:
I've done another implementation which seems to work equally well (also doesn't break the other tests!): https://github.com/pdm-project/pdm/compare/main...tgolsson:pdm:ts/1470-again. Let me know what 

I don't love the way I'm dealing with the group -> candidate info translation, but it seemed like the clearest way of handling it. What I wanted to do was generate a "synthetic" candidate + candidate info, but it seemed like more work than this approach, and I wasn't sure how it'd handle extras since those aren't known until we try to resolve.

Reply:
There may exist dynamic variables in the dependencies, so you should use `project.meta.dependencies` and `project.meta.optional_dependencies`, respectively.

And in L460:
```python
        if strip_extras(k)[0] == project.name:
            continue
```

`k` is normalized while `project.name` isn't. You need to normalize it with `normalize_name` and be careful of the possible `None` value.